### PR TITLE
Refactor nodes compatibility checking implement

### PR DESCRIPTION
**Pull request by slobber**
The original implement is focused on TPU compatibility only. Tensorflow Lite and Tensorflow.js also have the similar requirements to check the compatibility of their respective ops.
So our team refactor the implement to support more checkers to visualize the compatibility for any model type.

* Motivation for features / changes
The original implement is focused on TPU compatibility only. Tensorflow Lite and Tensorflow.js also have the similar requirements to check the compatibility of their respective ops.
So our team refactor the implement to support more checkers to visualize the compatibility for any model type.

* Technical description of changes
     - Design an interface of compatibility providers which has a function named `opValid` to check whether the node is compatible or not. Implement the TPU provider for current graph plugin.
    - Add a `node-title` in `tf-graph-op-compat-card` to support customizing the different titles of model targets.

* Screenshots of UI changes
No change.

* Detailed steps to verify changes work correctly (as executed by you)
`bazel run //tensorboard -- --logdir [log-path]`

* Alternate designs / implementations considered
None.

### CODE REVIEW

File: tensorboard/plugins/graph/tf_graph_dashboard/tf-graph-dashboard.html [hunk 1]
```diff
@@ -316,6 +318,9 @@ <h3>No graph definition files were found.</h3>
       // Either this dashboard is already initialized ... or we are not yet ready to initialize.
       return;
     }
+
+    this.set('_compatibilityProvider', new tf.graph.op.TpuCompatibilityProvider());
```
Comment (stephanwlee) [inline tensorboard/plugins/graph/tf_graph_dashboard/tf-graph-dashboard.html]: Create an explicit Polymer property for _compatibilityProvider

File: tensorboard/plugins/graph/tf_graph_dashboard/tf-graph-dashboard.html [hunk 2]
```diff
@@ -90,6 +90,7 @@ <h3>No graph definition files were found.</h3>
           out-stats="{{_stats}}"
           progress="{{_progress}}"
           out-hierarchy-params="{{_hierarchyParams}}"
+          compatibility-provider="{{_compatibilityProvider}}"
```
Comment (stephanwlee) [inline tensorboard/plugins/graph/tf_graph_dashboard/tf-graph-dashboard.html]: Please use square brackets unless tf-graph-loader notifies which it does not.
```suggestion
          compatibility-provider="[[_compatibilityProvider]]"
```

File: tensorboard/plugins/graph/tf_graph_info/tf-graph-info.html [hunk 3]
```diff
@@ -59,7 +59,8 @@
   <tf-graph-op-compat-card graph-hierarchy="[[graphHierarchy]]"
                            hierarchy-params="[[hierarchyParams]]"
                            render-hierarchy="[[renderHierarchy]]"
-                           color-by="[[colorBy]]">
+                           color-by="[[colorBy]]"
+                           node-title="[[compatNodeTitle]]">
```
Comment (stephanwlee) [inline tensorboard/plugins/graph/tf_graph_info/tf-graph-info.html]: create an explicit property for compatNodeTitle.

File: tensorboard/plugins/graph/tf_graph_op_compat_card/tf-graph-op-compat-card.html [hunk 4]
```diff
@@ -111,7 +111,7 @@
         on-click="_toggleExpanded"
         class="toggle-button">
       </paper-icon-button>
-      <div class="node-name" id="nodetitle">TPU Compatibility</div>
+      <div class="node-name" id="nodetitle">{{nodeTitle}}</div>
```
Comment (stephanwlee) [inline tensorboard/plugins/graph/tf_graph_op_compat_card/tf-graph-op-compat-card.html]: I think this should be square brackets again.

File: tensorboard/plugins/graph/tf_graph_common/op.ts [hunk 5]
```diff
@@ -14,435 +14,438 @@ limitations under the License.
...
+      'MergeSummary',
+      'ScalarSummary',
+      'StatsAggregatorSummary',
+    ];
+
+    /**
+     * Returns true if the node's inferred device is not the TPU.
+     * Note that this is only a best-effort check.
+     */
+    isNotTpuOp(opDevice: string): boolean {
```
Comment (stephanwlee) [inline tensorboard/plugins/graph/tf_graph_common/op.ts]: please prefix this with `private`. Though it used to be exported, it is not used anywhere else.

File: tensorboard/plugins/graph/tf_graph_common/op.ts [hunk 6]
```diff
@@ -14,435 +14,438 @@ limitations under the License.
...
+      }
+      // If assigned a device that is not TPU-related assume op is valid.
+      if (opNode.device && this.isNotTpuOp(opNode.device)) {
+        return true;
+      }
+      // If assigned to the TPU_SYSTEM device, assume op is valid.
+      if (opNode.device && opNode.device.search('TPU_SYSTEM') != -1) {
+        return true;
+      }
+      return TpuCompatibilityProvider.WHITELIST.indexOf(opNode.op) != -1;
```
Comment (stephanwlee) [inline tensorboard/plugins/graph/tf_graph_common/op.ts]: `Array.prototype.includes` is more semantically correct than `Array.prototype.indexOf`. Please prefer `includes`.

File: tensorboard/plugins/graph/tf_graph_common/op.ts [hunk 7]
```diff
@@ -14,435 +14,438 @@ limitations under the License.
...
     }
-    return WHITELIST.indexOf(opNode.op) != -1;
   }
 
-  export function checkOpsForCompatibility(graph: SlimGraph) {
+  export function checkOpsForCompatibility(graph: SlimGraph, provider: CompatibilityProvider) {
```
Comment (stephanwlee) [inline tensorboard/plugins/graph/tf_graph_common/op.ts]: nit: line-length

File: tensorboard/plugins/graph/tf_graph_common/op.ts [hunk 8]
```diff
@@ -14,435 +14,438 @@ limitations under the License.
...
+    if(provider === null) {
```
Comment (stephanwlee) [inline tensorboard/plugins/graph/tf_graph_common/op.ts]: Please add whitespace after `if` for consistency in the file.
```suggestion
    if (provider === null) {
```

File: tensorboard/plugins/graph/tf_graph_common/op.ts (hunk 6 above)
Comment (slobber) [reply tensorboard/plugins/graph/tf_graph_common/op.ts]: The `Array.prootype.includes` is defined in ES2016(ES7), but the tensorboard BUILD is using es6, so we can't use `includes` directly. 
See also:
https://github.com/tensorflow/tensorboard/blob/d285c7de73a09e9ece5a13368a58c00e1dcabda9/tensorboard/plugins/graph/tf_graph_common/op.ts#L432

Comment (slobber) [reply tensorboard/plugins/graph/tf_graph_common/op.ts]: I suppose `_.includes(collection, value)` from lodash lib.

File: tensorboard/plugins/graph/tf_graph_dashboard/tf-graph-dashboard.html [hunk 9]
```diff
@@ -316,6 +319,9 @@ <h3>No graph definition files were found.</h3>
       // Either this dashboard is already initialized ... or we are not yet ready to initialize.
       return;
     }
+
+    this.set('_compatibilityProvider', new tf.graph.op.TpuCompatibilityProvider());
```
[the remark this answers is not in the document]
Comment (slobber) [reply tensorboard/plugins/graph/tf_graph_dashboard/tf-graph-dashboard.html]: Sure, actually I have another pr that implements a tflite compatibility ops provider which ops is fetched from ajax, that provider must set the value in `_maybeInitializeDashboard`. For consistency, here also set the value in the observer.

**Comment by slobber** [comment]
> What's your thoughts on providing name of the compat-node-title via the compat-provider?
> It makes more sense conceptually but i18n can be a bit trickier (since DOM representation of datum is embedded in datum)

Setting compat-provider may be in an async process, if compat-node-title is gotten through the compat-provider, we should create a deeply broadcast from tf-graph-dashboard to tf-graph-op-compat-card. In view of the current state, I think it will only contain one provider in one plugin. So leaving a simple string property would be easier.

**Comment by stephanwlee** [comment]
> Setting compat-provider may be in an async process, if compat-node-title is gotten through the compat-provider, we should create a deeply broadcast from tf-graph-dashboard to tf-graph-op-compat-card. In view of the current state,

Don't see ripple of prop changes as a bad thing in an example especially since you said "one provider in one plugin".

Aside from this discussion, are you comfortable merging the PR as is?

**Comment by slobber** [comment]
> > Setting compat-provider may be in an async process, if compat-node-title is gotten through the compat-provider, we should create a deeply broadcast from tf-graph-dashboard to tf-graph-op-compat-card. In view of the current state,
> 
> Don't see ripple of prop changes as a bad thing in an example especially since you said "one provider in one plugin".
> 
> Aside from this discussion, are you comfortable merging the PR as is?

Sure, please merge this PR.

**Pull request merged**